### ISSUE TITLE
subrequest cycle problem

### ISSUE THREAD

**Issue by rubyist1982** [openresty/lua-nginx-module#2041]
nginx version: openresty/1.19.9.1
ngx_lua-0.10.20

-- my nginx.conf
worker_processes 1;
user root;
pcre_jit on;

events {
    worker_connections 1024;
}

http {
    lua_package_path '$prefix/?.lua;$prefix/resty_modules/lualib/?.lua;;';
    lua_code_cache off;
    default_type text/plain;
    keepalive_timeout 65;

    server {
        listen 8080;
        location / {
            content_by_lua_file lua/not_found.lua;
        }

        location = /main {
            content_by_lua_file lua/main.lua;
        }

        location = /sub {
            content_by_lua_file lua/sub.lua;
        }

    }
}

-- main.lua
local res = ngx.location.capture("/sub")
ngx.say(res.body)
ngx.say("main")

-- sub.lua
ngx.say("sub")

-- question
with lua_code_cache off;
when first issue curl /sub then curl /main
just got subrequest cycle error as follow

stack traceback:
coroutine 0:
        [C]: in function 'capture'
        /root/study/test/lua/main.lua:1: in main chunk, client: 127.0.0.1, server: , request: "GET /main HTTP/1.1", subrequest: "/sub", host: "localhost:8080"
2022/04/19 01:21:07 [error] 25279#25279: *10 lua subrequests cycle while processing "/sub", client: 127.0.0.1, server: , request: "GET /main HTTP/1.1", subrequest: "/sub", host: "localhost:8080"
2022/04/19 01:21:07 [error] 25279#25279: *10 lua entry thread aborted: runtime error: /root/study/test/lua/main.lua:1: failed to issue subrequest: -1
stack traceback:
coroutine 0:
        [C]: in function 'capture'
        /root/study/test/lua/main.lua:1: in main chunk, client: 127.0.0.1, server: , request: "GET /main HTTP/1.1", subrequest: "/sub", host: "localhost:8080"

and curl /main just print 51 line "mian" without any "sub"
is this intended ?
and i couldn't figure it out

**Comment by zhuizhuhaomeng** [comment]
```
-- sub.lua
local res = ngx.location.capture("/sub")
ngx.say(res.body)
ngx.say("main")
```
It it a typo?   ngx.say("main") should be  ngx.say("sub")

**Comment by rubyist1982** [comment]
sorry
-- sub.lua
ngx.say("sub")

> ```
> -- sub.lua
> local res = ngx.location.capture("/sub")
> ngx.say(res.body)
> ngx.say("main")
> ```
> 
> It it a typo? ngx.say("main") should be ngx.say("sub")

**Comment by rubyist1982** [comment]
> ```
> -- sub.lua
> local res = ngx.location.capture("/sub")
> ngx.say(res.body)
> ngx.say("main")
> ```
> 
> It it a typo? ngx.say("main") should be ngx.say("sub")

sorry for that. now i fix the typo.

**Comment by zhuizhuhaomeng** [comment]
Do you still have doubts?
you also remove the ngx.capture from sub.lua,  is that expected?

-- sub.lua
ngx.say("sub")
```

**Comment by rubyist1982** [comment]
> Do you still have doubts? you also remove the ngx.capture from sub.lua, is that expected?
> 
> ```
> -- sub.lua
> ngx.say("sub")
> ```

yes it is just that simple. and when lua_code_cache off;  it do produce subrquests cycle problem.

**Comment by gavin0616** [comment]
It seems that this issue is caused by the wrong value of `content_src_ref`. 
The `/sub` location retrieves an incorrect code chunk due to a wrong `content_src_ref`.

In this case:
1. **Request** /sub
- before:
   `/sub` `content_src_ref`: `LUA_REFNIL`
- after:
  `/sub` `content_src_ref`: `1`

2. **Request** /main
- before:
  `/main` `content_src_ref`: `LUA_REFNIL`
  `/sub`   `content_src_ref`: 1
- excute main.lua:
    1. Push `/main` code chunk to the top of lua stack
    2.   set `/main` `content_src_ref`: `1`
    3. Call `/sub` via `ngx.location.capture("/sub")`
        Attempt to use `/sub`'s `content_src_ref` to retrieve its code chunk. However, since `/sub`'s `content_src_ref` is the same as `/main`'s `content_src_ref`, the lookup mistakenly retrieves `/main`'s code chunk instead of actually reading the /`sub.lua` file. This causes `/main`'s code chunk to be executed recursively until it eventually results in an error.


Maybe we need to set *_src_ref to `LUA_REFNIL` before calling `ngx_http_lua_cache_load_code` when code_cache is disabled.

**Comment by gavin0616** [comment]
@zhuizhuhaomeng, could you please provide some feedback when you have time, at your convenience?
